src/coreclr/jit/llvmlower.cpp

```diff
@@ -929,40 +895,51 @@ void Llvm::lowerCallToShadowStack(GenTreeCall* callNode)
 
             if (corInfoType == CORINFO_TYPE_VALUECLASS)
             {
-                shadowStackUseOffest = padNextOffset(corInfoType, clsHnd, shadowStackUseOffest);
+                shadowStackUseOffset = padNextOffset(corInfoType, clsHnd, shadowStackUseOffset);
             }
             else
             {
-                shadowStackUseOffest += TARGET_POINTER_SIZE;
+                shadowStackUseOffset += TARGET_POINTER_SIZE;
             }
+
+            callNode->gtArgs.RemoveAfter(lastLlvmStackArg);
         }
         else
         {
             // Arg on LLVM stack.
             GenTreePutArgType* putArg = _compiler->gtNewPutArgType(argNode, corInfoType, clsHnd);
+            callArg.SetEarlyNode(putArg);
 #if DEBUG
-            putArg->SetArgNum(opAndArg.argNum);
+            putArg->SetArgNum(argIx);
 #endif
-            if (lastArg == nullptr)
-            {
-                lastArg = _compiler->gtNewCallArgs(putArg);
-                callNode->gtCallArgs = lastArg;
-            }
-            else
-            {
-                lastArg = _compiler->gtInsertNewCallArgAfter(putArg, lastArg);
-            }
-
             CurrentRange().InsertBefore(callNode, putArg);
-        }
-
-        if (isSigArg && (sigInfo != nullptr))
-        {
-            sigArgs = _info.compCompHnd->getArgNext(sigArgs);
+            lastLlvmStackArg = &callArg;
         }
 
         argIx++;
     }
+
+    callNode->gtArgs.MoveLateToEarly();
+
+    // Insert special args after finished iterating >gtArgs.Args
```

Comment:
To be honest, I liked the setup where we were inserting these in-order at the top more. But it's not a big deal.

Reply:
I just assumed it was going to be a problem with iterator.  I could do the iteration manually, starting from the last of shadow stack or return slot.  Do you want to see what that looks like?

Reply:
>  Do you want to see what that looks like?

It would be appreciated.

I don't think manual `GetNext`-style iteration is a big downside, since we already effectively depends on the argument list being linked with `RemoveAfter`.

Reply:
Last commit has this.  There's some long compound ways to get the head of the list `callNode->gtArgs.Args().begin().GetArg()`.  I could add a `GetHead` method to make that look nicer?

Reply:
Not worth it at present I think, the use in return lowering could be replaced with `GetArgByIndex(0)`, and then only only one is left, modulo upstream ones.

Reply:
>  I think, the use in return lowering could be replaced with GetArgByIndex(0)

I added a commit for this.  If it goes green I will ping Jan.  Thanks for the review!